importing libraries

In [3]:
import pandas as pd
import numpy as np

In [ ]:
# loading the dataset

customer_transaction_data = pd.read_csv(r'C:\Users\oluwa\OneDrive\Desktop\Fraudulent_transaction_Detection_for_a_FinTech_Company\Finlora_Dataset\FinLora_Customer_Transaction_Dataset.csv')
print("Shape:", customer_transaction_data.shape)

Shape: (11400, 26)


In [6]:
customer_transaction_data.head(10)

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.00,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.00,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.00,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.00,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.00,0
5,ec17d6f2-6505-4de9-910f-0de3ac3bbc23,4cfe5cbc-d677-49bc-8739-b1d05f351250,2022-10-04 12:09:59.468549+00:00,US,USD,GBP,mobile,526.9,526.90,8.75,...,0.094,low,616,0.702,0,0.361,0,0,0.00,0
6,2d8efe37-33ea-40de-9881-e553e3f2f8ce,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-04 12:37:41.468549+00:00,CA,CAD,GBP,mobile,149.24,110.44,2.33,...,0.299,standard,947,0.625,0,0.268,0,0,0.00,0
7,4d311532-ee0b-4985-b996-63e11a7144d3,4dfeffe2-94ed-4fc3-b5bc-89c6f1e86dab,2022-10-04 16:27:44.468549+00:00,CA,CAD,INR,web,276.51,204.62,5.59,...,0.087,standard,1041,0.867,0,0.245,0,1,0.12,0
8,db5bdea7-0701-4c07-9efa-815fa6ca9ede,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-04 21:00:36.468549+00:00,US,USD,EUR,mobile,99.52,99.52,2.41,...,0.182,enhanced,367,0.939,0,0.176,0,0,0.00,0
9,d74c81c0-bf5d-4184-8ce2-27fb753cb1c6,f7531a78-8bbe-4a85-b1e8-a0650edddc72,2022-10-04 21:20:42.468549+00:00,US,USD,PHP,web,302.55,302.55,5.26,...,0.413,standard,1016,0.944,0,0.247,0,0,0.10,1


In [ ]:
# columns in the dataset

customer_transaction_data.columns

Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

In [ ]:
# checking data types and null counts in the dataset
print("DTYPES:")
print(customer_transaction_data.dtypes)
print()
print("NULL COUNTS:")
print(customer_transaction_data.isnull().sum())

DTYPES:
transaction_id                object
customer_id                   object
timestamp                     object
home_country                  object
source_currency               object
dest_currency                 object
channel                       object
amount_src                    object
amount_usd                   float64
fee                          float64
exchange_rate_src_to_dest    float64
device_id                     object
new_device                      bool
ip_address                    object
ip_country                    object
location_mismatch               bool
ip_risk_score                float64
kyc_tier                      object
account_age_days               int64
device_trust_score           float64
chargeback_history_count       int64
risk_score_internal          float64
txn_velocity_1h                int64
txn_velocity_24h               int64
corridor_risk                float64
is_fraud                       int64
dtype: object

NULL COUNTS:
tr

There are 11,400 rows and 26 columns. Most column data typed as expected (numbers as float64, int64, flags as bool) but amount_src typed as text (object) instead of numeric. Many columns have nulls: timestamp (29), amount_usd (305), fee (295), ip_address (305), ip_country (301), kyc_tier (300), device_trust_score (295)

### Data Dictionary

transaction_id — Unique reference for the transaction.

customer_id — Reference for the customer who made the transaction.

timestamp — Date and time the transaction occurred (UTC).

home_country — Customer's registered/home country.

source_currency — Currency the transaction was sent in.

dest_currency — Currency the transaction was converted to / sent to.

channel — Channel used to initiate the transaction (mobile / web / ATM).

amount_src — Transaction amount in the source currency.

amount_usd — Transaction amount converted to USD.

fee — Fee charged on the transaction.

exchange_rate_src_to_dest — Exchange rate applied between source and destination currency.

device_id — Device used to make the transaction.

new_device — Whether this is the first transaction from this device for this customer.

ip_address — IP address the transaction originated from.

ip_country — Country resolved from the IP address.

location_mismatch — Whether ip_country differs from home_country.

ip_risk_score — Risk score associated with the originating IP address.

kyc_tier — Customer's Know-Your-Customer verification tier.

account_age_days — Age of the customer's account in days at time of transaction.

device_trust_score — Trust score assigned to the device (0–1 scale).

chargeback_history_count — Number of prior chargebacks on the customer's account.

risk_score_internal — Finlora's internal composite risk score for the transaction.

txn_velocity_1h — Number of transactions by the customer in the last hour.

txn_velocity_24h — Number of transactions by the customer in the last 24 hours.

corridor_risk — Risk rating associated with the source-to-destination currency corridor.

is_fraud — Confirmed fraud outcome (1 = fraud, 0 = legitimate).

In [ ]:
# amount_src show as text when it should be float. Checking the unique values in the column to see if there are any non-numeric values.

def is_num(x):
    try:
        float(x)
        return True
    except:
        return False

mask = ~customer_transaction_data['amount_src'].apply(is_num)
print("Non-numeric-looking values:", mask.sum())
print(customer_transaction_data.loc[mask, 'amount_src'].unique())

Non-numeric-looking values: 4
['9,998.85' '9,995.95' '1,541.55' '9,991.24']


There are four values that have comma thousands-separators making pandas unable to parse them as numbers thus making the whole column text instead.

In [11]:
for c in customer_transaction_data.columns:
    print(c, customer_transaction_data[c].nunique())

transaction_id 11200
customer_id 1315
timestamp 11141
home_country 7
source_currency 3
dest_currency 9
channel 12
amount_src 9856
amount_usd 9621
fee 1676
exchange_rate_src_to_dest 25
device_id 2113
new_device 2
ip_address 10900
ip_country 9
location_mismatch 2
ip_risk_score 952
kyc_tier 14
account_age_days 412
device_trust_score 677
chargeback_history_count 3
risk_score_internal 651
txn_velocity_1h 10
txn_velocity_24h 12
corridor_risk 7
is_fraud 2


transaction_id should be a unique reference but it has 11,200 unique values across 11,400 rows — fewer than the row count, meaning some transaction IDs are likely diplicated. Also kyc_tier shows 14 unique values despite it being a 3 to 4 category field. channel also shows 12 for what should be 3 entries. These findings indicate messy category labels.

In [13]:
# check duplicate transaction_id values and duplicate rows

dupe_ids = customer_transaction_data['transaction_id'][customer_transaction_data['transaction_id'].duplicated(keep=False)]
print("Rows sharing a duplicated transaction_id:", dupe_ids.shape[0])

print("Fully duplicate rows (every column identical):", customer_transaction_data.duplicated().sum())

Rows sharing a duplicated transaction_id: 400
Fully duplicate rows (every column identical): 200


400 rows share a transaction_id with another row (i.e. 200 transaction_ids appear twice each), and exactly 200 rows are complete duplicates across every column. That means all of these repeats are simple accidental double-entries, not conflicting records - probably safe to drop.

In [ ]:
# checking categorical columns for consistency, unique values and their counts

for col in ['kyc_tier', 'channel', 'home_country', 'ip_country']:
    print(f"--- {col} ---")
    print(customer_transaction_data[col].value_counts(dropna=False))
    print()

--- kyc_tier ---
kyc_tier
standard       7931
enhanced       1829
low            1039
NaN             300
standrd          72
STANDARD         70
 standard        65
unknown          32
 enhanced        17
ENHANCED         17
enhancd          12
 low              8
LOW               5
 nan              2
NAN               1
Name: count, dtype: int64

--- channel ---
channel
mobile       6366
web          3727
ATM          1008
mobille        60
 mobile        48
MOBILE         47
unknown        37
WEB            36
 web           34
weeb           24
ATm             9
 ATM            4
Name: count, dtype: int64

--- home_country ---
home_country
US         7940
UK         2119
CA         1221
 US          68
unknown      32
 UK          13
 CA           7
Name: count, dtype: int64

--- ip_country ---
ip_country
US         6848
UK         2407
CA         1726
NaN         301
 US          49
unknown      33
 UK          18
 CA          15
NAN           2
 nan          1
Name: count, dtyp

Each of these fields have the categories buried under inconsistent casing, stray whitespace, typos, and placeholder strings - e.g. channel is really just mobile / web / ATM, but appears as 12 raw variants (" mobile", "MOBILE", "mobille", "weeb", etc. ); kyc_tier is really standard / enhanced / low but appears as 14 variants including literal "unknown", "nan", "NAN" strings that are not actually NaN. This needs to be standardised (lowercase, strip whitespace, map typos) before any grouping or modelling can be done.

In [15]:
# checking source_currency and dest_currency for the same kind of issue.

print(customer_transaction_data['source_currency'].value_counts())
print()
print(customer_transaction_data['dest_currency'].value_counts())

source_currency
USD    8031
GBP    2139
CAD    1230
Name: count, dtype: int64

dest_currency
NGN    1474
USD    1328
INR    1325
CNY    1324
PHP    1314
GBP    1296
CAD    1282
EUR    1196
MXN     861
Name: count, dtype: int64


Both columns seem clean. There are 3 source currencies (USD, GBP, CAD) and 9 destination currencies, no casing/typo issues noticed here. 

In [23]:
# confirming target variable — is_fraud is binary and checking the class balance.

print(customer_transaction_data['is_fraud'].value_counts())
print(customer_transaction_data['is_fraud'].value_counts(normalize=True).reset_index(name='proportion'))

is_fraud
0    10403
1      997
Name: count, dtype: int64
   is_fraud  proportion
0         0    0.912544
1         1    0.087456


is_fraud is binary (0/1), there are no missing values. 10,403 legitimate transactions (91.25%) vs 997 fraud transactions (8.75%) — a meaningfully class imbalance that will require handling later.

In [ ]:
# Descriptive statistics of the numeric fields, to spot obviously wrong ranges or negative values where none should exist

num_cols = ['amount_usd', 'fee', 'exchange_rate_src_to_dest', 'account_age_days',
            'device_trust_score', 'chargeback_history_count', 'risk_score_internal',
            'txn_velocity_1h', 'txn_velocity_24h', 'corridor_risk', 'ip_risk_score']
print(customer_transaction_data[num_cols].describe().T)

                             count        mean          std    min      25%  \
amount_usd                 11095.0  452.022083  1403.973062  7.230   92.465   
fee                        11105.0  100.309441   958.128504 -1.000    2.380   
exchange_rate_src_to_dest  11400.0  167.540397   382.023827  0.592    1.000   
account_age_days           11400.0  393.793158   342.348393  1.000  147.000   
device_trust_score         11105.0    0.653681     0.273012 -0.100    0.515   
chargeback_history_count   11400.0    0.048509     0.256194  0.000    0.000   
risk_score_internal        11400.0    0.267134     0.142983  0.000    0.169   
txn_velocity_1h            11400.0    0.458333     1.524494 -1.000    0.000   
txn_velocity_24h           11400.0    0.723509     1.958390  0.000    0.000   
corridor_risk              11400.0    0.045501     0.084942  0.000    0.000   
ip_risk_score              11400.0    0.396726     0.270507  0.004    0.209   

                                  50%         75%  

Most fields look okay. However, fee stands out having a minimum of -1 and a maximum of 9999.99 (fees can't sensibly be negative or about £10k on everyday transfers), and ip_risk_score has a maximum of 1.2, which is odd if the score is meant to be between 0 and 1. txn_velocity_1h has a minimum of -1 and you can't have made negative transactions in the last hour.
device_trust_score has a minimum of -0.1 — everything else in that column is between 0 and 1.

In [ ]:
# checking the distribution of transactions among the customers, to see if there are any customers with unusually high number of transactions.

cust_counts = customer_transaction_data['customer_id'].value_counts()
print(cust_counts.describe())
print("Customers with exactly 1 transaction:", (cust_counts == 1).sum(), "/ total customers:", len(cust_counts))
print("Largest customer:", cust_counts.index[0], "with", cust_counts.iloc[0], "transactions")

count    1315.000000
mean        8.669202
std        84.766961
min         1.000000
25%         1.000000
50%         1.000000
75%         1.000000
max      1510.000000
Name: count, dtype: float64
Customers with exactly 1 transaction: 1239 / total customers: 1315
Largest customer: 402cccc9-28de-45b3-9af7-cc5302aa1f93 with 1510 transactions


In 11400 records, there are 1315 unique customers and 1239 of them (94%) have only one transaction. One of the customer_id entries appears to be an outlier since it has 1510 transactions, which comprise 13.2% of all the dataset. It needs to be checked if it is legitimately a high volume account or data artifact.

In [26]:
missing_ip_address = customer_transaction_data[customer_transaction_data['ip_address'].isnull()]
print(missing_ip_address.shape)
missing_ip_address

(305, 26)


,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
13,1a34a095-3be4-4a69-a22e-803be566526c,7041b9c1-3719-4ca8-9a6b-811b47cea6c0,2022-10-05 07:04:39.468549+00:00,UK,GBP,EUR,mobile,322.24,NaN,NaN,...,0.245,NaN,4,NaN,0,0.407,0,0,0.05,0
49,b489474c-b729-4759-86e9-51778ddea385,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-08 20:45:48.468549+00:00,US,USD,INR,mobile,255.06,NaN,NaN,...,0.530,NaN,263,NaN,0,0.223,0,0,0.00,0
151,32cbd613-5c6c-4797-be15-165bc408c2ca,af8ca4c4-8703-4c55-b66c-2b76cd70040d,2022-10-20 22:08:59.468549+00:00,US,USD,MXN,mobile,502.97,NaN,NaN,...,0.310,NaN,1018,NaN,0,0.087,0,0,0.20,0
163,8b339135-7446-4689-9bd2-0509f23903f0,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-22 01:51:23.468549+00:00,CA,CAD,USD,web,100.87,NaN,NaN,...,0.519,NaN,947,NaN,0,0.268,7,7,0.00,0
168,c6c7f296-be97-4d18-a93e-719fcba43dce,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-22 10:08:19.468549+00:00,US,USD,MXN,mobile,265.3,NaN,NaN,...,0.655,NaN,367,NaN,0,0.176,0,0,0.20,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10054,35da0a3e-5776-4f7e-b102-5ea3d5034713,402cccc9-28de-45b3-9af7-cc5302aa1f93,2025-06-20 07:38:53.468549+00:00,US,USD,EUR,mobile,9993.94,NaN,NaN,...,0.434,NaN,263,NaN,0,0.223,0,0,0.00,0
10082,adc74103-d3b7-491e-a3c9-02aae2050d46,7bd5200c-5d19-44f0-9afe-8b339a05366b,2025-01-27 17:14:38.468549+00:00,US,USD,USD,mobile,414.29,NaN,NaN,...,0.446,NaN,147,NaN,0,0.391,0,0,0.00,1
10093,d5ce08b7-d78f-4e1b-97c0-362cc64e12ee,af8ca4c4-8703-4c55-b66c-2b76cd70040d,2024-06-28 06:10:20.468549+00:00,US,USD,PHP,mobile,125.75,NaN,NaN,...,0.451,NaN,1018,NaN,0,0.087,0,0,0.10,0
10142,356a1a21-e76c-4ad6-b712-405d3ad215e6,23d9943d-57b1-42eb-9ef2-05eed7fd1957,2024-03-04 21:11:40.468549+00:00,UK,GBP,INR,mobile,83.96,NaN,NaN,...,0.474,NaN,718,NaN,0,0.169,0,0,0.00,0


In [28]:
missing_ip_address.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                      0
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     0
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

In [29]:
print(missing_ip_address['is_fraud'].value_counts(normalize=True).reset_index(name='proportion'))

   is_fraud  proportion
0         0     0.97377
1         1     0.02623


##### Among the 305 rows where the ip_addresses are missing, 97.4% of the transactions are legitimate and only 2.6% are fraud. Missing rows are disproportionately non-fraud, and not a random sample of the data.

#### Summary/Observations
The Dataset contains 11,400 rows and 26 columns, with a single record representing a transaction.Several issues were discovered while exploring the data that require further attention before any modelling can take place:

1. amount_src column is of type object instead of numeric due to four instances of values with comma-thousand separators which pandas cannot parse correctly

2. transaction_id field is not unique, with 200 transaction_ids appearing twice, and all of those pairs being completely identical rows, so it is safe to remove those with drop_duplicates()

3. Categorical columns contain numerous typos and inconsistencies, namely:

channel has 12 unique values when there should only be 3 (the mobile/web/ATM categories)

kyc_tier has 14 unique values, of which 3-4 are supposed to be different tiers, while the rest are either incorrectly labelled low, medium, or high, or are "unknown"/"nan"/"NAN" case variants, indicating a lack of information

home_country and ip_country also have mixed casing and spacing which needs to be standardized. All other categorical columns are fine, with source_currency/dest_currency having no issues

4. is_fraud is the target variable, it's correctly typed as binary, has no missing values, but has an extreme class imbalance of 91.25% / 8.75%

5. Some numeric fields have values that don't make any sense in context such as fee ranging from -1 to nearly 10 000, ip_risk_score going up to 1.2 when it should be between 0 and 1, txn_velocity_1h's minimum of -1, and device_trust_score's minimum of -0.1, which suggests those were placeholder values

6. There is one customer_id that stands out exceptionally, with 1 510 of the 11 400 transactions belonging to them, compared to 94% of all customers having only 1 transaction each, which needs cross-verifying with the business before attempting any customer-level analysis

### DATA CLEANING AND PREPARATION

In [100]:


customer_transaction_data = customer_transaction_data.drop_duplicates().reset_index(drop=True)

In [101]:
customer_transaction_data.shape

(10836, 26)

In [102]:
# converting datatype of timestamp to datetime format

customer_transaction_data['timestamp'] = pd.to_datetime(customer_transaction_data['timestamp'], errors='coerce')
# converting datatype of income_usd to float

customer_transaction_data['amount_src'] = pd.to_numeric(customer_transaction_data['amount_src'], errors='coerce')

In [103]:
# checking to see data types have been converted successfully.

customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10836 entries, 0 to 10835
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             10836 non-null  object             
 1   customer_id                10836 non-null  object             
 2   timestamp                  10836 non-null  datetime64[ns, UTC]
 3   home_country               10836 non-null  object             
 4   source_currency            10836 non-null  object             
 5   dest_currency              10836 non-null  object             
 6   channel                    10836 non-null  object             
 7   amount_src                 10836 non-null  float64            
 8   amount_usd                 10836 non-null  float64            
 9   fee                        10836 non-null  float64            
 10  exchange_rate_src_to_dest  10836 non-null  float64            
 11  de

In [104]:
# Rechecking for null values
customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

In [105]:
# Getting the exchange rate for the currencies in the dataset. 

exchange_rate = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(
    lambda x: (x['amount_usd'] / x['amount_src']).mean()).to_dict()
exchange_rate

C:\Users\oluwa\AppData\Local\Temp\ipykernel_964\2038650960.py:3: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  exchange_rate = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(


{'CAD': 0.7225170084075527,
 'GBP': 1.224162125960103,
 'USD': 0.9837248982806143}

In [106]:
# Filling the missing values in the amount_usd column using the exchange rate for the source_currency.
customer_transaction_data['amount_usd'] = customer_transaction_data.apply(
    lambda row: row['amount_usd'] if pd.notna(row['amount_usd']) else row['amount_src'] * exchange_rate.get(row['source_currency'], 1),
      axis=1
      )

In [107]:
# Standardizing the categorical columns(home_country, ip_country, kyc_tier, channel) to have consistent values. 

# Channel 

customer_transaction_data['channel'] = customer_transaction_data['channel'].str.strip().str.capitalize()

customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({
    'Mobille': 'Mobile',
    'Weeb': 'Web',
})

# Home_country

customer_transaction_data['home_country'] = customer_transaction_data['home_country'].str.strip()

# ip_country

customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].str.strip()
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].replace(
    {'NAN': np.nan, 'nan': np.nan}
)
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna('unknown')
print(customer_transaction_data['ip_country'].value_counts(dropna=False))

# kyc tier

customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].str.strip().str.lower()
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace(
    {'nan': np.nan}
)
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({
    'standrd': 'standard',
    'enhancd': 'enhanced',
})
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].fillna('unknown')
print(customer_transaction_data['kyc_tier'].value_counts(dropna=False))



ip_country
US         6724
UK         2375
CA         1707
unknown      30
Name: count, dtype: int64
kyc_tier
standard    7937
enhanced    1828
low         1043
unknown       28
Name: count, dtype: int64


In [108]:
customer_transaction_data['kyc_tier'].value_counts(dropna=False)

kyc_tier
standard    7937
enhanced    1828
low         1043
unknown       28
Name: count, dtype: int64

In [109]:
# filling missing values in the fee column using the median fee for the source_currency.

if 'fee' in customer_transaction_data.columns:
        if 'channel' in customer_transaction_data.columns:
                customer_transaction_data['fee'] = customer_transaction_data.groupby('channel')['fee'].transform(lambda x: x.fillna(x.median()))
                customer_transaction_data['fee'] = customer_transaction_data['fee'].fillna(customer_transaction_data['fee'].median())


In [110]:
# filling missing values in the ip_country column with corresponding values from the home_country column

if {'ip_country', 'home_country'}.issubset(customer_transaction_data.columns):
    customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna(customer_transaction_data['home_country'])

In [111]:
# filling missing values in the device_trust_score column using the median device_trust_score per group of new device and kyc_tier if they exist Otherwise, they would  with the overall median.

if 'device_trust_score' in customer_transaction_data.columns:
    if {'new device', 'kyc_tier'}.issubset(customer_transaction_data.columns):
        customer_transaction_data['device_trust_score'] = customer_transaction_data.groupby(['new device', 'kyc_tier'])['device_trust_score'].transform(lambda x: x.fillna(x.median()))
    customer_transaction_data['device_trust_score'] = customer_transaction_data['device_trust_score'].fillna(customer_transaction_data['device_trust_score'].median())

In [112]:
# rechecking for null values

customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

In [113]:
missing_timestamp = customer_transaction_data[customer_transaction_data['timestamp'].isna()]
print("Rows with missing/invalid timestamp:", missing_timestamp.shape[0])
print("is_fraud rate among these rows:", missing_timestamp['is_fraud'].mean())
print("Overall is_fraud rate:", customer_transaction_data['is_fraud'].mean())

Rows with missing/invalid timestamp: 0
is_fraud rate among these rows: nan
Overall is_fraud rate: 0.09099298634182355


In [114]:
missing_ip_address = customer_transaction_data[customer_transaction_data['ip_address'].isna()]
print("Rows with missing/invalid IP address:", missing_ip_address.shape[0])
print("is_fraud rate among these rows:", missing_ip_address['is_fraud'].mean())
print("Overall is_fraud rate:", customer_transaction_data['is_fraud'].mean())

Rows with missing/invalid IP address: 0
is_fraud rate among these rows: nan
Overall is_fraud rate: 0.09099298634182355


In [115]:
# dropping ip_address and timestamp columns as they cannot be filled and the percentage of missing values is somewhat low

customer_transaction_data = customer_transaction_data.dropna(subset=['timestamp', 'amount_src', 'ip_address'])

print(customer_transaction_data.shape)
print(customer_transaction_data['is_fraud'].mean())

(10836, 26)
0.09099298634182355


In [116]:
# dropping the columns that were created to indicate missing values in the ip_address, timestamp, and kyc_tier columns, as they are no longer needed after dropping the rows with missing values in those columns.

customer_transaction_data = customer_transaction_data.drop(['ip_missing', 'timestamp_missing', 'kyc_tier_missing'], axis=1)

KeyError: "['ip_missing', 'timestamp_missing', 'kyc_tier_missing'] not found in axis"

In [ ]:
#rechecking for null values after dropping the rows with missing timestamp and ip_address

customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
kyc_tier_missing             0
timestamp_missing            0
ip_missing                   0
dtype: int64

In [ ]:
print("Exact duplicate rows:", customer_transaction_data.duplicated().sum())
print("Duplicate transaction_id values:", customer_transaction_data['transaction_id'].duplicated().sum())

Exact duplicate rows: 194
Duplicate transaction_id values: 194


In [ ]:
dupes = customer_transaction_data[customer_transaction_data.duplicated(keep=False)].sort_values('transaction_id')
dupes.head(10)

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
10153,03368cab-696b-4984-bffb-3429267e57b2,7bd5200c-5d19-44f0-9afe-8b339a05366b,2024-06-29 04:15:54.468549+00:00,US,USD,GBP,Atm,299.24,299.24,4.70,...,0.202,standard,147,0.551,0,0.391,0,0,0.0,0
5866,03368cab-696b-4984-bffb-3429267e57b2,7bd5200c-5d19-44f0-9afe-8b339a05366b,2024-06-29 04:15:54.468549+00:00,US,USD,GBP,Atm,299.24,299.24,4.70,...,0.202,standard,147,0.551,0,0.391,0,0,0.0,0
6264,0368fc3c-2323-470e-aafe-eb0d920553eb,d71c91b4-fee8-4104-9856-a5c6109a62e3,2024-08-16 14:03:44.468549+00:00,US,USD,EUR,Mobile,118.40,118.40,2.82,...,0.154,standard,298,0.336,0,0.166,0,0,0.0,0
10079,0368fc3c-2323-470e-aafe-eb0d920553eb,d71c91b4-fee8-4104-9856-a5c6109a62e3,2024-08-16 14:03:44.468549+00:00,US,USD,EUR,Mobile,118.40,118.40,2.82,...,0.154,standard,298,0.336,0,0.166,0,0,0.0,0
10036,04532993-f980-4c87-9129-373a00ed57db,d71c91b4-fee8-4104-9856-a5c6109a62e3,2024-04-18 10:26:16.468549+00:00,US,USD,PHP,Mobile,438.81,438.81,6.98,...,0.310,standard,298,0.336,0,0.166,0,0,0.1,0
5185,04532993-f980-4c87-9129-373a00ed57db,d71c91b4-fee8-4104-9856-a5c6109a62e3,2024-04-18 10:26:16.468549+00:00,US,USD,PHP,Mobile,438.81,438.81,6.98,...,0.310,standard,298,0.336,0,0.166,0,0,0.1,0
6700,061c3f56-bff6-47d9-b195-1ecdc1faee8a,af8ca4c4-8703-4c55-b66c-2b76cd70040d,2024-10-06 05:17:22.468549+00:00,US,USD,MXN,Mobile,82.11,82.11,1.95,...,0.385,standard,1018,0.934,0,0.087,0,0,0.2,0
10085,061c3f56-bff6-47d9-b195-1ecdc1faee8a,af8ca4c4-8703-4c55-b66c-2b76cd70040d,2024-10-06 05:17:22.468549+00:00,US,USD,MXN,Mobile,82.11,82.11,1.95,...,0.385,standard,1018,0.934,0,0.087,0,0,0.2,0
2572,068abac2-c06a-42af-9a38-379886b26e4a,d71c91b4-fee8-4104-9856-a5c6109a62e3,2023-07-14 03:58:42.468549+00:00,US,USD,USD,Mobile,91.31,91.31,0.96,...,0.362,standard,298,0.336,0,0.166,0,0,0.0,0
10136,068abac2-c06a-42af-9a38-379886b26e4a,d71c91b4-fee8-4104-9856-a5c6109a62e3,2023-07-14 03:58:42.468549+00:00,US,USD,USD,Mobile,91.31,91.31,0.96,...,0.362,standard,298,0.336,0,0.166,0,0,0.0,0


In [ ]:
num_cols = ['amount_src', 'amount_usd', 'fee', 'device_trust_score', 'risk_score_internal',
            'txn_velocity_1h', 'txn_velocity_24h']
negative_counts = {col: (customer_transaction_data[col] < 0).sum() for col in num_cols}

print("Negative values in each column:")
for col, count in negative_counts.items():
    print(f"{col}: {count}")

Negative values in each column:
amount_src: 100
amount_usd: 0
fee: 90
device_trust_score: 190
risk_score_internal: 0
txn_velocity_1h: 190
txn_velocity_24h: 0


In [ ]:


negative_counts = {(col, (customer_transaction_data[col] < 0).sum()) for col in num_cols}